include/blast/math/panel/Gemm.hpp

```diff
@@ -5,105 +5,76 @@
-            BLAZE_THROW_INVALID_ARGUMENT("Matrix sizes do not match");
-
-        size_t i = 0;
+            BLAST_THROW_EXCEPTION(std::invalid_argument {"Matrix sizes do not match"});
 
-        // i + 4 * PANEL_SIZE != M is to improve performance in case when the remaining number of rows is 4 * PANEL_SIZE:
-        // it is more efficient to apply 2 * PANEL_SIZE kernel 2 times than 3 * PANEL_SIZE + 1 * PANEL_SIZE kernel.
-        for (; i + 2 * SS < M && i + 4 * SS != M; i += 3 * SS)
-            gemm_nt_backend<3 * SS, TILE_STEP>(M, N, K, i, alpha, ptr(*A), ~trans(ptr(*B)), beta, ptr(*C), ptr(*D));
-
-        for (; i + 1 * SS < M; i += 2 * SS)
-            gemm_nt_backend<2 * SS, TILE_STEP>(M, N, K, i, alpha, ptr(*A), ~trans(ptr(*B)), beta, ptr(*C), ptr(*D));
-
-        for (; i + 0 * SS < M; i += 1 * SS)
-            gemm_nt_backend<1 * SS, TILE_STEP>(M, N, K, i, alpha, ptr(*A), ~trans(ptr(*B)), beta, ptr(*C), ptr(*D));
+        gemm(M, N, K, alpha, ptr(*A), ptr(*B), beta, ptr(*C), ptr(*D));
     }
 
 
+    /**
+     * @brief Matrix-matrix multiplication for @a PanelMatrix arguments
+     *
+     * D := A*B + C
+     *
+     * A, B and C are matrices, with A
+     * an m by k matrix, B a k by n matrix and C an m by n matrix.
+     *
+     * @param A the matrix A
+     * @param B the matrix B
+     * @param C the matrix C
+     * @param D the output matrix D
+     */
     template <
-        size_t KM, size_t KN, typename T,
-        typename MPA, typename MPB, typename MPC, typename MPD
+        typename MT1, typename MT2, bool SO2,
+        typename MT3, typename MT4
     >
-    requires MatrixPointer<MPA, T> && MatrixPointer<MPB, T> && MatrixPointer<MPC, T> && MatrixPointer<MPD, T>
-    BLAZE_ALWAYS_INLINE void gemm_nt_backend(size_t M, size_t N, size_t K, size_t i, T alpha, MPA A, MPB B, T beta, MPC C, MPD D)
+    inline void gemm(
+        PanelMatrix<MT1, columnMajor> const& A, PanelMatrix<MT2, SO2> const& B,
+        PanelMatrix<MT3, columnMajor> const& C, PanelMatrix<MT4, columnMajor>& D)
     {
-        using ET = ElementType_t<MPD>;
-        RegisterMatrix<ET, KM, KN, columnMajor> ker;
-
-        if (i + KM <= M)
-        {
-            size_t j = 0;
-            auto a = A(i, 0);
-
-            for (; j + KN <= N; j += KN)
-                gemm(ker, K, alpha, a, B(0, j), beta, C(i, j), D(i, j));
-
-            if (j < N)
-                gemm(ker, K, alpha, a, B(0, j), beta, C(i, j), D(i, j), KM, N - j);
-        }
-        else
-        {
-            // Use partial save to calculate the bottom of the resulting matrix.
-            size_t j = 0;
-
-            for (; j + KN <= N; j += KN)
-                gemm(ker, K, alpha, A(i, 0), B(0, j), beta, C(i, j), D(i, j), M - i, KN);
-
-            if (j < N)
-                gemm(ker, K, alpha, A(i, 0), B(0, j), beta, C(i, j), D(i, j), M - i, N - j);
-        }
+        using ET = ElementType_t<MT4>;
+        gemm(ET(1.), A, B, ET(1.), C, D);
```

Comment:
Awesome. But I'm missing something: was it already implemented in `gemm`? 

Or did you get rid of a special case that didn't make sense anymore? Or both?

Reply:
Yes, I got rid of a special case.

Reply:
:+1: